analyzers/tests/SonarAnalyzer.TestFramework/Verification/IssueValidation/IssueLocationPair.cs

```diff
@@ -45,7 +47,7 @@ public void AppendAssertionMessage(StringBuilder builder)
         {
             builder.Append(" Rule ").Append(Actual.RuleId);
         }
-        if ((Actual?.IssueId ?? Expected?.IssueId) is { } issueId)
+        if ((Actual?.IssueId ?? Expected?.IssueId) is { } issueId && Actual?.RuleId != issueId)
```

Comment:
If I am not mistaken, this only applies to build errors.
From a reader's POV, it seems weird to compare ruleId with issueId.
Could you leave a comment explaining it?

Something similar to:
```csharp
// Error [CS1001] should assert that expected issueId matches the actual RuleId
```

...from the `IssueLocation` class. 
I think it's fine to have it commented in two places, since it's a very special-case behavior.

Reply:
I'll change this in #8671, as the problem will likely disappear there